internal/controller/nonadminrestore_controller.go

```diff
@@ -0,0 +1,339 @@
+				Status:  metav1.ConditionFalse,
+				Reason:  "InvalidRestoreSpec",
+				Message: err.Error(),
+			},
+		)
+		if updatedPhase || updatedCondition {
+			if updateErr := r.Status().Update(ctx, nar); updateErr != nil {
+				logger.Error(updateErr, statusUpdateError)
+				return false, updateErr
+			}
+		}
+		return false, reconcile.TerminalError(err)
+	}
+	logger.V(1).Info("NonAdminRestore Spec validated")
+
+	updated := meta.SetStatusCondition(&nar.Status.Conditions,
+		metav1.Condition{
+			Type:    string(constant.NonAdminConditionAccepted),
+			Status:  metav1.ConditionTrue,
+			Reason:  "RestoreAccepted",
+			Message: "restore accepted",
+		},
+	)
+	if updated {
+		if err := r.Status().Update(ctx, nar); err != nil {
+			logger.Error(err, statusUpdateError)
+			return false, err
+		}
+		logger.V(1).Info("NonAdminRestore condition set to Accepted")
+	}
+	return false, nil
+}
+
+func (r *NonAdminRestoreReconciler) setUUID(ctx context.Context, logger logr.Logger, nar *nacv1alpha1.NonAdminRestore) (bool, error) {
+	if nar.Status.UUID == constant.EmptyString {
+		// TODO handle panic
+		nar.Status.UUID = uuid.New().String()
+		if err := r.Status().Update(ctx, nar); err != nil {
+			logger.Error(err, statusUpdateError)
+			return false, err
+		}
+		logger.V(1).Info(veleroReferenceUpdated)
+	} else {
+		logger.V(1).Info("NonAdminRestore already contains NAC UUID")
+	}
+	return false, nil
+}
+
+func (r *NonAdminRestoreReconciler) setFinalizer(ctx context.Context, logger logr.Logger, nar *nacv1alpha1.NonAdminRestore) (bool, error) {
+	added := controllerutil.AddFinalizer(nar, constant.NonAdminRestoreFinalizerName)
+	if added {
+		if err := r.Update(ctx, nar); err != nil {
+			logger.Error(err, "Failed to add finalizer")
+			return false, err
+		}
+		// TODO does this change generation? need to refetch?
+		logger.V(1).Info("Finalizer added to NonAdminRestore", "finalizer", constant.NonAdminRestoreFinalizerName)
+	} else {
+		logger.V(1).Info("Finalizer already exists on NonAdminRestore", "finalizer", constant.NonAdminRestoreFinalizerName)
+	}
+	return false, nil
+}
+
+func (r *NonAdminRestoreReconciler) createVeleroRestore(ctx context.Context, logger logr.Logger, nar *nacv1alpha1.NonAdminRestore) (bool, error) {
+	nacUUID := nar.Status.UUID
+	veleroRestore, err := function.GetVeleroRestoreByLabel(ctx, r.Client, r.OADPNamespace, nacUUID)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return false, err
+		}
+
+		// TODO how to add this to validation without doing GET call for NAB/Velero backup twice? add field in status?
+		restoreSpec := nar.Spec.RestoreSpec.DeepCopy()
+		nab := &nacv1alpha1.NonAdminBackup{}
+		err = r.Get(ctx, types.NamespacedName{Namespace: nar.Namespace, Name: nar.Spec.RestoreSpec.BackupName}, nab)
+		if err != nil {
+			return false, err
+		}
+		// TODO better way to check readiness? simplify and ask user to pass velero backup name? (user has access to this info in nonAdminBackup status)
```

Comment:
We should maybe requeue until it's ready for restore?

Reply:
this may never happen (example after 1 hour in progress, velero restore fails). Would try something more immediate here